src/app/qgsrelationmanagerdialog.cpp

```diff
@@ -61,47 +103,101 @@ void QgsRelationManagerDialog::addRelation( const QgsRelation &rel )
-  mRelationsTable->setItem( row, 4, item );
+  // Save relation in first column's item
+  item->setData( 0, Qt::UserRole, QVariant::fromValue<QgsRelation>( rel ) );
 
-  item = new QTableWidgetItem( rel.id() );
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 5, item );
+  item->setText( 0, rel.name() );
+  item->setText( 1, rel.referencedLayer()->name() );
+  item->setText( 2, referencedFields );
+  item->setText( 3, rel.referencingLayer()->name() );
+  item->setText( 4, referencingFields );
+  item->setText( 5, rel.id() );
+  item->setText( 6, rel.strength() == QgsRelation::RelationStrength::Composition
+                 ? QStringLiteral( "Composition" )
+                 : QStringLiteral( "Association" ) );
 
-  if ( rel.strength() == QgsRelation::RelationStrength::Composition )
+  mRelationsTree->setSortingEnabled( true );
+}
+
+void QgsRelationManagerDialog::addPolymorphicRelation( const QgsPolymorphicRelation &polyRel )
+{
+  if ( ! polyRel.isValid() )
+    return;
+
+  QString referencingFields;
+  QString referencedFields;
+
+  for ( int i = 0; i < polyRel.fieldPairs().count(); i++ )
   {
-    item = new QTableWidgetItem( QStringLiteral( "Composition" ) );
+    if ( i != 0 )
+    {
+      referencingFields += QStringLiteral( ", " );
+      referencedFields += QStringLiteral( ", " );
+    }
+
+    referencingFields += polyRel.fieldPairs().at( i ).referencingField();
+    referencedFields += polyRel.fieldPairs().at( i ).referencedField();
   }
-  else
+
+  mRelationsTree->setSortingEnabled( false );
+  int row = mRelationsTree->topLevelItemCount();
+  QTreeWidgetItem *item = new QTreeWidgetItem();
+
+  mRelationsTree->insertTopLevelItem( row, item );
+
+  // Save relation in first column's item
+  item->setExpanded( true );
+  item->setFlags( item->flags() | Qt::ItemIsEditable );
+  item->setData( 0, Qt::UserRole, QVariant::fromValue<QgsPolymorphicRelation>( polyRel ) );
+  item->setText( 0, polyRel.name() );
+  item->setText( 1, polyRel.referencingLayer()->name() );
+  item->setText( 2, referencedFields );
+  item->setText( 3, QStringLiteral( "as in \"%1\".\"%2\"" ).arg( polyRel.referencingLayer()->name(), polyRel.referencedLayerField() ) );
+  item->setText( 4, referencingFields );
+  item->setText( 5, polyRel.id() );
+  item->setText( 6, polyRel.strength() == QgsRelation::RelationStrength::Composition
+                 ? QStringLiteral( "Composition" )
+                 : QStringLiteral( "Association" ) );
+
+  const QList<QgsRelation> generatedRelations = polyRel.generateRelations();
+  for ( const QgsRelation &generatedRelation : generatedRelations )
   {
-    item = new QTableWidgetItem( QStringLiteral( "Association" ) );
+    if ( !generatedRelation.isValid() )
+      continue;
+
+    addRelation( generatedRelation );
   }
-  item->setFlags( Qt::ItemIsEnabled );
-  mRelationsTable->setItem( row, 6, item );
 
-  mRelationsTable->setSortingEnabled( true );
+  mRelationsTree->setSortingEnabled( true );
```

Comment:
need to be called each time?

Reply:
I guess it is, check the commit msg 31a1c23